// bmad-dashboard/server/artifactDetails.js
import path from "path";
import fs from "fs";
import { projectRoot } from "./config.js";
import {
  extractFrontmatterAndBody,
  parseMarkdownSections,
} from "./fileUtils.js";
import {
  PRD_STEP_LABELS,
  PRD_STEP_DESCRIPTIONS,
  ARCH_STEP_IDS,
  ARCH_STEP_LABELS,
  ARCH_STEP_DESCRIPTIONS,
} from "./constants.js";

const MAX_SCAN_DEPTH = 10;

/** Match AC line: bold or plain Given/When/Then/And; support - and * bullets */
const AC_LINE_BOLD =
  /^[-*]\s*\*\*(Given|When|Then|And)\s*:?\s*(?:\*\*\s*)?(.+)$/;
const AC_LINE_PLAIN = /^[-*]\s*(Given|When|Then|And)\s*:?\s*(.+)$/;

function parseStoryBody(body, num, title, statusByKey) {
  const id = num.replace(".", "-");
  const rest = (body || "").replace(/\r\n/g, "\n").replace(/\r/g, "\n").trim();
  let userStory = "";
  const acceptanceCriteria = [];
  const acMarkers = [
    /\*\*Acceptance Criteria\s*:?\s*\*\*/i,
    /\*\*Acceptance criteria\s*:?\s*\*\*/i,
    /^Acceptance Criteria\s*:?\s*$/im,
    /^Acceptance criteria\s*:?\s*$/im,
    /^#+\s*Acceptance Criteria\s*:?\s*$/im,
    /^#+\s*Acceptance criteria\s*:?\s*$/im,
    /^Acceptance Criteria\s*:?\s*\n/im,
    /\bAC\s*:?\s*\n/im,
  ];
  let acIndex = -1;
  let acBlockStart = 0;
  for (const re of acMarkers) {
    const m = rest.match(re);
    if (m) {
      acIndex = m.index;
      acBlockStart = m.index + m[0].length;
      break;
    }
  }
  if (acIndex >= 0) {
    const beforeAc = rest.slice(0, acIndex).trim();
    const paras = beforeAc.split(/\n\n+/).filter((p) => p.trim().length > 15);
    const para = paras.find(
      (p) =>
        /As a[\s\S]+I want[\s\S]+so that/.test(p.replace(/\n/g, " ")) ||
        p.trim().length > 30,
    );
    if (para) userStory = para.replace(/\n/g, " ").trim();
    const acBlock = rest.slice(acBlockStart).trim();
    const acLines = acBlock
      .split(/\n/)
      .map((line) => line.trim())
      .filter(Boolean);
    for (const line of acLines) {
      if (line.startsWith("###")) break;
      let m = line.match(AC_LINE_BOLD);
      if (!m) m = line.match(AC_LINE_PLAIN);
      if (m) acceptanceCriteria.push({ type: m[1], text: m[2].trim() });
      else if (
        (line.startsWith("- ") || line.startsWith("* ")) &&
        line.length > 2
      )
        acceptanceCriteria.push({ type: null, text: line.slice(2).trim() });
      else {
        const numBullet = line.match(/^\d+\.\s+(.+)$/);
        if (numBullet && numBullet[1].trim().length > 0)
          acceptanceCriteria.push({ type: null, text: numBullet[1].trim() });
      }
    }
    if (acceptanceCriteria.length === 0 && acBlock.length > 0) {
      for (const line of acLines) {
        if (line.startsWith("###")) break;
        const bullet =
          line.match(/^[-*]\s+(.+)$/) || line.match(/^\d+\.\s+(.+)$/);
        if (bullet && bullet[1].trim().length > 0)
          acceptanceCriteria.push({ type: null, text: bullet[1].trim() });
      }
    }
  }
  if (acceptanceCriteria.length === 0 && rest) {
    const lines = rest.split(/\n/);
    for (const line of lines) {
      const t = line.trim();
      if (t.startsWith("###")) continue;
      let m = t.match(AC_LINE_BOLD);
      if (!m) m = t.match(AC_LINE_PLAIN);
      if (m) acceptanceCriteria.push({ type: m[1], text: m[2].trim() });
      else if ((t.startsWith("- ") || t.startsWith("* ")) && t.length > 2)
        acceptanceCriteria.push({ type: null, text: t.slice(2).trim() });
      else {
        const numBullet = t.match(/^\d+\.\s+(.+)$/);
        if (numBullet && numBullet[1].trim().length > 0)
          acceptanceCriteria.push({ type: null, text: numBullet[1].trim() });
      }
    }
  }

  if (!userStory && rest) {
    const first = rest.split(/\n\n+/)[0];
    if (first && first.length > 10)
      userStory = first.replace(/\n/g, " ").trim();
  }
  const statusKey = Object.keys(statusByKey).find(
    (k) => k === id || k.startsWith(id + "-"),
  );
  const status = statusKey ? statusByKey[statusKey] : null;
  const progress =
    status === "done"
      ? 100
      : status === "in-progress" || status === "review"
        ? 50
        : 0;
  return {
    num,
    id,
    title,
    userStory: userStory || null,
    acceptanceCriteria: acceptanceCriteria.length ? acceptanceCriteria : null,
    status,
    progress,
  };
}

const ANALYSIS_ARTIFACT_NAMES = new Set([
  "prd.md",
  "architecture.md",
  "epics.md",
]);

async function getAnalysisArtifacts(planPath, outputRoot, rel) {
  const items = [];
  if (!planPath || !fs.existsSync(planPath)) return items;

  const outputDir =
    outputRoot && fs.existsSync(outputRoot) ? outputRoot : planPath;

  async function addFile(filePath, type, label) {
    if (!fs.existsSync(filePath)) return;
    try {
      const { body } = await extractFrontmatterAndBody(filePath);
      const sections = parseMarkdownSections(body);
      items.push({
        type,
        label,
        sourcePath: rel(filePath),
        sections: sections.length
          ? sections
          : [
              {
                title: "Content",
                body: body.trim().slice(0, 100000) || "(No content)",
              },
            ],
      });
    } catch (e) {
      console.error("addFile failed:", e);
    }
  }

  const brainstormDir = path.join(outputDir, "brainstorming");
  if (fs.existsSync(brainstormDir)) {
    try {
      if (!fs.statSync(brainstormDir).isDirectory())
        throw new Error("not a dir");
      const entries = await fs.promises.readdir(brainstormDir, {
        withFileTypes: true,
      });
      for (const e of entries) {
        if (e.isFile() && e.name.endsWith(".md"))
          await addFile(
            path.join(brainstormDir, e.name),
            "brainstorm",
            `Brainstorm: ${e.name}`,
          );
      }
    } catch (e) {
      console.error("getAnalysisArtifacts brainstormDir failed:", e);
    }
  }

  const researchDir = path.join(planPath, "research");
  async function scanResearch(dir, depth = 0) {
    if (depth > MAX_SCAN_DEPTH) return;
    let entries;
    try {
      entries = await fs.promises.readdir(dir, { withFileTypes: true });
    } catch (e) {
      console.error("scanResearch readdir failed:", e);
      return;
    }
    for (const e of entries) {
      const full = path.join(dir, e.name);
      if (e.isFile() && e.name.endsWith(".md"))
        await addFile(full, "research", `Research: ${e.name}`);
      else if (e.isDirectory()) await scanResearch(full, depth + 1);
    }
  }
  if (fs.existsSync(researchDir)) {
    try {
      if (!fs.statSync(researchDir).isDirectory()) throw new Error("not a dir");
      await scanResearch(researchDir);
    } catch (e) {
      console.error("getAnalysisArtifacts researchDir failed:", e);
    }
  }

  let planEntries = [];
  try {
    planEntries = await fs.promises.readdir(planPath, { withFileTypes: true });
  } catch (e) {
    console.error("getAnalysisArtifacts readdir failed:", e);
  }
  for (const e of planEntries) {
    if (!e.isFile() || !e.name.endsWith(".md")) continue;
    if (ANALYSIS_ARTIFACT_NAMES.has(e.name)) continue;
    const lower = e.name.toLowerCase();
    if (lower.includes("research"))
      await addFile(path.join(planPath, e.name), "research", e.name);
    else if (lower.includes("brief") || lower.includes("product-brief"))
      await addFile(path.join(planPath, e.name), "product-brief", e.name);
    else if (lower.includes("brainstorm"))
      await addFile(path.join(planPath, e.name), "brainstorm", e.name);
  }

  return items;
}

export async function getArtifactDetails(
  projectRootParam,
  planPath,
  implPath,
  sprintData,
) {
  const root = projectRootParam ?? projectRoot;
  const details = {
    prd: null,
    architecture: null,
    epics: null,
    sprintStatus: null,
    analysis: null,
  };
  const rel = (p) =>
    root && p ? path.relative(root, p).replace(/\\/g, "/") : "";

  if (!planPath || !fs.existsSync(planPath)) return details;

  const outputRoot = planPath ? path.dirname(planPath) : null;
  const analysisItems = await getAnalysisArtifacts(planPath, outputRoot, rel);
  if (analysisItems.length > 0) {
    details.analysis = {
      summary: `${analysisItems.length} analysis artifact(s): brainstorm, research, and/or product brief`,
      items: analysisItems,
    };
  }

  const prdPath = path.join(planPath, "prd.md");
  if (fs.existsSync(prdPath)) {
    const { frontmatter: fm, body: prdBody } =
      await extractFrontmatterAndBody(prdPath);
    const completedRaw =
      fm && Array.isArray(fm.stepsCompleted) ? fm.stepsCompleted : [];
    const completedSet = new Set(
      completedRaw.map((x) => String(x).trim()).filter(Boolean),
    );
    const steps = Object.entries(PRD_STEP_LABELS).map(([id, label]) => ({
      id,
      label,
      description: PRD_STEP_DESCRIPTIONS[id] || "",
      completed: completedSet.has(id),
    }));
    const completedCount = steps.filter((s) => s.completed).length;
    details.prd = {
      sourcePath: rel(prdPath),
      commands: ["/bmad-bmm-create-prd"],
      completedAt: fm && fm.date ? fm.date : null,
      summary: `${completedCount} of ${steps.length} steps completed`,
      steps,
      classification: fm && fm.classification ? fm.classification : null,
      sections: parseMarkdownSections(prdBody),
    };
  }

  const archPath = path.join(planPath, "architecture.md");
  if (fs.existsSync(archPath)) {
    const { frontmatter: fm, body: archBody } =
      await extractFrontmatterAndBody(archPath);
    const completedRaw =
      fm && Array.isArray(fm.stepsCompleted) ? fm.stepsCompleted : [];
    const archCompletedSet = new Set(
      completedRaw.map((x) => String(x).trim()).filter(Boolean),
    );
    const archSteps = ARCH_STEP_IDS.map((n) => {
      const completed = archCompletedSet.has(String(n));
      return {
        id: String(n),
        label: ARCH_STEP_LABELS[n],
        description: ARCH_STEP_DESCRIPTIONS[n] || "",
        completed,
      };
    });
    const archDoneCount = archSteps.filter((s) => s.completed).length;
    details.architecture = {
      sourcePath: rel(archPath),
      commands: ["/bmad-bmm-create-architecture"],
      completedAt: fm && fm.completedAt ? fm.completedAt : null,
      summary: `${archDoneCount} of ${archSteps.length} steps completed`,
      steps: archSteps,
      sections: parseMarkdownSections(archBody),
    };
  }

  const epicsPath = path.join(planPath, "epics.md");
  if (fs.existsSync(epicsPath)) {
    const { body: content } = await extractFrontmatterAndBody(epicsPath);
    const epics = [];
    const epicRe = /^## Epic (\d+):\s*(.+)$/gm;
    const epicBlocks = content.split(/^## Epic \d+:/m).slice(1);
    const epicTitles = [...content.matchAll(epicRe)].map((e) => ({
      num: e[1],
      title: e[2].trim(),
    }));
    const statusByKey =
      sprintData && sprintData.development_status
        ? sprintData.development_status
        : {};

    for (let i = 0; i < epicTitles.length; i++) {
      const block = epicBlocks[i] || "";
      const storyRe = /^### Story (\d+)\.(\d+)\s*[:-]\s*(.+)$/gm;
      const storyMatches = [...block.matchAll(storyRe)];
      const rawBodies = block.split(/^### Story \d+\.\d+\s*[:-]\s*/m).slice(1);
      const stories = storyMatches.map((m, j) => {
        const num = `${m[1]}.${m[2]}`;
        const title = m[3].trim();
        const body = rawBodies[j] || "";
        return parseStoryBody(body, num, title, statusByKey);
      });
      epics.push({
        epicNum: epicTitles[i].num,
        title: epicTitles[i].title,
        stories,
      });
    }
    details.epics = {
      sourcePath: rel(epicsPath),
      commands: [
        "/bmad-bmm-create-epics-and-stories",
        "/bmad-bmm-sprint-planning",
      ],
      summary: `${epics.length} epics with stories (sprint plan from sprint-status).`,
      list: epics,
    };
  }

  if (sprintData && sprintData.development_status) {
    const status = sprintData.development_status;
    const items = Object.entries(status)
      .filter(([k]) => !k.includes("retrospective"))
      .map(([id, value]) => ({ id, status: value }));
    const sprintPath = implPath
      ? path.join(implPath, "sprint-status.yaml")
      : "";
    details.sprintStatus = {
      sourcePath: rel(sprintPath),
      commands: [
        "/bmad-bmm-sprint-planning",
        "/bmad-bmm-create-story",
        "/bmad-bmm-dev-story",
        "/bmad-bmm-code-review",
        "/bmad-bmm-retrospective",
      ],
      summary:
        "Tracks story/epic status for implementation. Run create-story → dev-story → code-review per story.",
      generated: sprintData.generated || null,
      items,
      epics:
        details.epics && details.epics.list
          ? details.epics.list.map((e) => ({
              id: `epic-${e.epicNum}`,
              epicNum: e.epicNum,
              title: e.title,
              status: status[`epic-${e.epicNum}`] || null,
              stories: e.stories.map((s) => ({
                id: s.id,
                num: s.num,
                title: s.title,
                status: s.status,
                progress: s.progress,
                userStory: s.userStory,
                acceptanceCriteria: s.acceptanceCriteria,
              })),
            }))
          : null,
    };
  }

  return details;
}
