// bmad-dashboard/server/bmadState.js
import path from 'path'
import fs from 'fs'
import { bmadRoot, projectRoot } from './config.js'
import { loadYaml, resolvePath, hasArtifact, findFiles } from './fileUtils.js'
import { PHASE_ORDER, ARTIFACT_KEYS } from './constants.js'

/** Resolve one file path per artifact key for mtime check; returns { [key]: path } */
function getArtifactPaths(bmm) {
  const plan = bmm.planning_artifacts
  const impl = bmm.implementation_artifacts
  const out = {}
  if (plan && fs.existsSync(plan)) {
    const prd = findFiles(plan, 'prd.md')[0] || findFiles(plan, '*prd*.md')[0]
    if (prd) out.hasPrd = prd
    const arch = findFiles(plan, 'architecture.md')[0] || findFiles(plan, '*architecture*.md')[0]
    if (arch) out.hasArchitecture = arch
    const epics = findFiles(plan, 'epics.md')[0] || findFiles(plan, '*epic*.md')[0]
    if (epics) out.hasEpics = epics
    const ux = findFiles(plan, '*ux*.md')[0]
    if (ux) out.hasUx = ux
  }
  if (impl && fs.existsSync(impl)) {
    const sprintPath = path.join(impl, 'sprint-status.yaml')
    if (fs.existsSync(sprintPath)) out.hasSprintStatus = sprintPath
  }
  return out
}

/** Which artifact key was last modified (by file mtime) — used to blink "last executed" command */
export function getLastExecutedArtifactKey(bmm) {
  const paths = getArtifactPaths(bmm)
  let latestKey = null
  let latestMtime = 0
  for (const key of ARTIFACT_KEYS) {
    const p = paths[key]
    if (!p) continue
    try {
      const stat = fs.statSync(p)
      if (stat.mtimeMs > latestMtime) {
        latestMtime = stat.mtimeMs
        latestKey = key
      }
    } catch {
      // ignore missing / permission
    }
  }
  return latestKey
}

export async function getBmmConfig() {
  const configPath = path.join(bmadRoot, 'bmm', 'config.yaml')
  const config = await loadYaml(configPath)
  if (!config) return {}
  return {
    planning_artifacts: resolvePath(config.planning_artifacts, projectRoot),
    implementation_artifacts: resolvePath(config.implementation_artifacts, projectRoot),
    project_name: config.project_name,
  }
}

export async function detectArtifacts(bmm) {
  const plan = bmm.planning_artifacts
  const impl = bmm.implementation_artifacts
  return {
    hasPrd: await hasArtifact(plan, ['prd.md', '*prd*.md']),
    hasArchitecture: await hasArtifact(plan, ['architecture.md', '*architecture*.md']),
    hasEpics: (await hasArtifact(plan, ['epics.md', '*epic*.md'])) || (plan && (await findFiles(plan, 'epic*.md')).length > 0),
    hasUx: await hasArtifact(plan, ['*ux*.md']),
    hasSprintStatus: impl && fs.existsSync(path.join(impl, 'sprint-status.yaml')),
    storyFiles: impl ? (await findFiles(impl, '*.md')).filter((f) => !f.includes('sprint-status') && !f.includes('retro')) : [],
  }
}

export function inferBmmPhase(artifacts, workflows) {
  const required = workflows.filter((w) => w.module === 'bmm' && w.required === 'true')
  const ordered = [...required].sort((a, b) => {
    const pa = PHASE_ORDER.indexOf(a.phase)
    const pb = PHASE_ORDER.indexOf(b.phase)
    if (pa !== pb) return pa - pb
    return (parseInt(a.sequence, 10) || 0) - (parseInt(b.sequence, 10) || 0)
  })

  for (let i = ordered.length - 1; i >= 0; i--) {
    const w = ordered[i]
    const code = (w.code || '').toLowerCase()
    if (code === 'cp' && artifacts.hasPrd) return { phase: '2-planning', nextSequence: 20 }
    if (code === 'cu' && artifacts.hasUx) continue
    if (code === 'ca' && artifacts.hasArchitecture) return { phase: '3-solutioning', nextSequence: 30 }
    if (code === 'ce' && artifacts.hasEpics) return { phase: '3-solutioning', nextSequence: 70 }
    if (code === 'ir' && artifacts.hasSprintStatus) return { phase: '4-implementation', nextSequence: 10 }
    if ((code === 'sp' || code === 'cs') && artifacts.hasSprintStatus) return { phase: '4-implementation', nextSequence: 30 }
  }

  if (artifacts.hasSprintStatus) return { phase: '4-implementation', nextSequence: null }
  if (artifacts.hasEpics) return { phase: '3-solutioning', nextSequence: 70 }
  if (artifacts.hasArchitecture) return { phase: '3-solutioning', nextSequence: 30 }
  if (artifacts.hasPrd) return { phase: '2-planning', nextSequence: 20 }
  return { phase: '1-analysis', nextSequence: 10 }
}

export async function getSprintStatus(implPath) {
  const file = path.join(implPath, 'sprint-status.yaml')
  if (!fs.existsSync(file)) return null
  const data = await loadYaml(file)
  if (!data || !data.development_status) return null
  const status = data.development_status
  const epics = []
  const stories = []
  let currentEpic = null
  for (const [key, value] of Object.entries(status)) {
    // Match top-level epic keys like "epic-1" but not "epic-1-story-2" or "epic-1-retro"
    if (/^epic-\d+$/.test(key)) {
      currentEpic = { id: key, status: value, stories: [] }
      epics.push(currentEpic)
    } else if (currentEpic && typeof value === 'string') {
      currentEpic.stories.push({ id: key, status: value })
    } else if (typeof value === 'string' && !key.includes('retro')) {
      stories.push({ id: key, status: value })
    }
  }
  if (epics.length === 0) {
    for (const [key, value] of Object.entries(status)) {
      if (typeof value === 'string' && !key.includes('retro')) stories.push({ id: key, status: value })
    }
  }
  return {
    project: data.project,
    generated: data.generated,
    development_status: status,
    epics,
    stories,
  }
}

export function buildWorkflowList(workflows) {
  return workflows
    .filter((w) => w.module === 'bmm' && w.phase !== 'anytime')
    .map((w) => ({
      phase: w.phase,
      sequence: w.sequence,
      name: w.name,
      code: w.code,
      command: w.command,
      required: w.required === 'true',
      agent: w['agent-display-name'],
      description: w.description,
    }))
}

export function buildNextSteps(workflows, phaseInfo, _artifacts) {
  const bmmWorkflows = workflows.filter((w) => w.module === 'bmm' && w.phase !== 'anytime')
  const currentPhaseIndex = PHASE_ORDER.indexOf(phaseInfo.phase)
  const next = []
  for (const w of bmmWorkflows) {
    const pi = PHASE_ORDER.indexOf(w.phase)
    if (pi < currentPhaseIndex) continue
    if (pi === currentPhaseIndex && phaseInfo.nextSequence != null && (parseInt(w.sequence, 10) || 0) < phaseInfo.nextSequence) continue
    if (w.command) next.push({ ...w, commandDisplay: `/bmad-${w.command.replace(/_/g, '-')}` })
    if (next.length >= 8) break
  }
  return next.slice(0, 6)
}
