// bmad-dashboard/server/constants.js
export const PHASE_ORDER = ['1-analysis', '2-planning', '3-solutioning', '4-implementation']

export const ARTIFACT_KEYS = ['hasPrd', 'hasArchitecture', 'hasEpics', 'hasUx', 'hasSprintStatus']

export const PRD_STEP_LABELS = {
  'step-01-init': '1. Init',
  'step-02-discovery': '2. Discovery',
  'step-02b-vision': '2b. Vision',
  'step-02c-executive-summary': '2c. Executive Summary',
  'step-03-success': '3. Success Criteria',
  'step-04-journeys': '4. User Journeys',
  'step-05-domain': '5. Domain',
  'step-06-innovation': '6. Innovation',
  'step-07-project-type': '7. Project Type',
  'step-08-scoping': '8. Scoping',
  'step-09-functional': '9. Functional Requirements',
  'step-10-nonfunctional': '10. Non-Functional Requirements',
  'step-11-polish': '11. Polish',
  'step-12-complete': '12. Complete',
}

export const PRD_STEP_DESCRIPTIONS = {
  'step-01-init': 'Set up PRD workspace, discover input docs, create document from template.',
  'step-02-discovery': 'Classify project type, domain, complexity, and context (greenfield/brownfield).',
  'step-02b-vision': 'Discover product vision and what makes it special.',
  'step-02c-executive-summary': 'Draft Executive Summary from discovery; append to PRD.',
  'step-03-success': 'Define user, business, and technical success criteria and scope (MVP/Growth/Vision).',
  'step-04-journeys': 'Map user types and narrative journeys.',
  'step-05-domain': 'Capture domain requirements.',
  'step-06-innovation': 'Capture innovation focus.',
  'step-07-project-type': 'Confirm project type and implications.',
  'step-08-scoping': 'Define MVP boundaries and prioritize features.',
  'step-09-functional': 'Synthesize functional requirements (capability contract).',
  'step-10-nonfunctional': 'Synthesize non-functional requirements.',
  'step-11-polish': 'Polish and finalize PRD.',
  'step-12-complete': 'Complete PRD workflow.',
}

export const ARCH_STEP_LABELS = {
  '1': '1. Init',
  '2': '2. Project Context',
  '3': '3. Starter / Stack',
  '4': '4. Core Decisions',
  '5': '5. Patterns',
  '6': '6. Project Structure',
  '7': '7. Validation',
  '8': '8. Complete',
}

export const ARCH_STEP_IDS = Object.keys(ARCH_STEP_LABELS)

export const ARCH_STEP_DESCRIPTIONS = {
  '1': 'Initialize architecture workflow; discover PRD and inputs.',
  '2': 'Analyze requirements and project context for architectural implications.',
  '3': 'Evaluate starter template / stack and document decisions it provides.',
  '4': 'Make core architectural decisions (data, auth, API, frontend, infra).',
  '5': 'Define implementation patterns and consistency rules for agents.',
  '6': 'Define project structure and map requirements to components.',
  '7': 'Validate architecture coherence and completeness.',
  '8': 'Complete workflow and hand off to implementation.',
}
