// bmad-dashboard/server/fileUtils.js
import path from "path";
import fs from "fs";
import { parse as parseYaml } from "yaml";

const FILE_SIZE_LIMIT_BYTES = 1024 * 1024; // 1 MB

export function resolvePath(raw, root) {
  if (!raw || typeof raw !== "string") return null;
  return raw.replace(/\{project-root\}/g, root);
}

async function readFileSafe(filePath) {
  try {
    const stat = await fs.promises.stat(filePath);
    if (stat.size > FILE_SIZE_LIMIT_BYTES) return null;
    return await fs.promises.readFile(filePath, "utf8");
  } catch {
    return null;
  }
}

export async function loadYaml(filePath) {
  const content = await readFileSafe(filePath);
  if (content === null) return null;
  try {
    return parseYaml(content);
  } catch {
    return null;
  }
}

export function parseCsvLine(line) {
  const out = [];
  let cur = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (c === '"') {
      if (inQuotes && line[i + 1] === '"') {
        // escaped double-quote inside a quoted field
        cur += '"';
        i++;
      } else {
        inQuotes = !inQuotes;
      }
    } else if (c === "," && !inQuotes) {
      out.push(cur.trim());
      cur = "";
    } else {
      cur += c;
    }
  }
  out.push(cur.trim());
  return out;
}

export async function loadCsv(filePath) {
  const content = await readFileSafe(filePath);
  if (content === null) return [];
  try {
    const lines = content.trim().split("\n");
    if (lines.length < 2) return [];
    const headers = parseCsvLine(lines[0]);
    return lines.slice(1).map((line) => {
      const values = parseCsvLine(line);
      const row = {};
      headers.forEach((h, i) => (row[h] = values[i] ?? ""));
      return row;
    });
  } catch {
    return [];
  }
}

export function patternToRe(glob) {
  // Escape all regex metacharacters before expanding glob wildcards
  const escaped = glob.replace(/[.+?^${}()|[\]\\]/g, "\\$&");
  const s = escaped
    .replace(/\*\*/g, "<<>>")
    .replace(/\*/g, "[^/]*")
    .replace(/<<>>/g, ".*");
  return new RegExp(`^${s}$`);
}

export async function findFiles(dir, pattern, recursive = false) {
  if (!dir) return [];
  try {
    await fs.promises.access(dir);
  } catch {
    return [];
  }
  const re = patternToRe(pattern);
  const out = [];
  async function scan(d) {
    let entries;
    try {
      entries = await fs.promises.readdir(d, { withFileTypes: true });
    } catch {
      return;
    }
    for (const e of entries) {
      const full = path.join(d, e.name);
      if (e.isFile() && re.test(e.name)) out.push(full);
      if (e.isDirectory() && recursive) await scan(full);
    }
  }
  await scan(dir);
  return out;
}

export async function hasArtifact(dir, patterns) {
  if (!dir) return false;
  let names;
  try {
    names = await fs.promises.readdir(dir);
  } catch {
    return false;
  }
  for (const p of patterns) {
    const re = patternToRe(p);
    if (names.some((n) => re.test(n))) return true;
  }
  return false;
}

export async function extractFrontmatter(filePath) {
  const raw = await readFileSafe(filePath);
  if (raw === null) return null;
  try {
    const match = raw.match(/^---\r?\n([\s\S]*?)\r?\n---/);
    if (!match) return null;
    return parseYaml(match[1]);
  } catch {
    return null;
  }
}

/** Returns { frontmatter, body } from a markdown file with YAML frontmatter. */
export async function extractFrontmatterAndBody(filePath) {
  const raw = await readFileSafe(filePath);
  if (raw === null) return { frontmatter: null, body: "" };
  try {
    const match = raw.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n([\s\S]*)$/);
    if (!match) return { frontmatter: null, body: raw };
    return {
      frontmatter: parseYaml(match[1]),
      body: (match[2] || "").trim(),
    };
  } catch {
    return { frontmatter: null, body: "" };
  }
}

/** Truncate at last newline before limit so markdown (e.g. tables) is not cut mid-line. */
function truncateAtLine(text, maxLen) {
  if (text.length <= maxLen) return text;
  const slice = text.slice(0, maxLen + 1);
  const lastNewline = slice.lastIndexOf("\n");
  const cut = lastNewline >= 0 ? lastNewline : maxLen;
  return text.slice(0, cut).trimEnd() + "…";
}

/** Parse markdown body into sections by ## or ### headings. Each section: { title, body }. Body trimmed; truncated only if over maxSectionLength (default high so full content is shown in UI). */
export function parseMarkdownSections(body, maxSectionLength = 500000) {
  if (!body || typeof body !== "string") return [];
  const sections = [];
  const re = /^(#{2,3})\s+(.+)$/gm;
  let match;
  let lastEnd = 0;
  let lastTitle = null;
  while ((match = re.exec(body)) !== null) {
    if (lastTitle != null) {
      const text = body.slice(lastEnd, match.index).trim();
      sections.push({
        title: lastTitle,
        body:
          text.length > maxSectionLength
            ? truncateAtLine(text, maxSectionLength)
            : text,
      });
    }
    lastTitle = match[2].trim();
    lastEnd = match.index + match[0].length;
  }
  if (lastTitle != null) {
    const text = body.slice(lastEnd).trim();
    sections.push({
      title: lastTitle,
      body:
        text.length > maxSectionLength
          ? truncateAtLine(text, maxSectionLength)
          : text,
    });
  }
  return sections.filter((s) => s.body.length > 0 || s.title.length > 0);
}
