// bmad-dashboard/server/fileWatch.js
import path from 'path'
import fs from 'fs'

const DEBOUNCE_MS = 400
let watcherCleanups = []
let debounceTimer = null
const subscribers = new Set()

function notify() {
  debounceTimer = null
  subscribers.forEach((send) => {
    try {
      send()
    } catch {
      subscribers.delete(send)
    }
  })
}

function debouncedNotify() {
  if (debounceTimer) clearTimeout(debounceTimer)
  debounceTimer = setTimeout(notify, DEBOUNCE_MS)
}

function watchDir(dir) {
  if (!dir || !fs.existsSync(dir)) return
  // Note: { recursive: true } is silently ignored on Linux < Node 22 (works only on macOS/Windows).
  // On affected Linux versions, only top-level changes in each watched directory are detected.
  try {
    const w = fs.watch(dir, { recursive: true }, (event, filename) => {
      if (filename) debouncedNotify()
    })
    watcherCleanups.push(() => {
      try { w.close() } catch (e) { console.error('watcher close failed:', e) }
    })
  } catch {
    // Fall back to non-recursive watch (recursive: true not supported on this platform)
    try {
      const w = fs.watch(dir, (event, filename) => {
        if (filename) debouncedNotify()
      })
      watcherCleanups.push(() => {
        try { w.close() } catch (e) { console.error('watcher close failed:', e) }
      })
    } catch (e) {
      console.error(`watchDir failed for ${dir}:`, e)
    }
  }
}

/**
 * Start watching project paths that affect BMAD state.
 * @param {string} projectRoot - Project root (parent of _bmad-output, _bmad)
 */
export function startWatching(projectRoot) {
  if (!projectRoot) return
  const outputPath = path.join(projectRoot, '_bmad-output')
  const bmadPath = path.join(projectRoot, '_bmad')
  const configPath = path.join(bmadPath, '_config')
  const bmmPath = path.join(bmadPath, 'bmm')

  watchDir(outputPath)
  watchDir(configPath)
  watchDir(bmmPath)
}

/**
 * Stop all active watchers and clear the debounce timer.
 * Resets internal state so startWatching can be called again safely.
 */
export function stopWatching() {
  if (debounceTimer) {
    clearTimeout(debounceTimer)
    debounceTimer = null
  }
  for (const cleanup of watcherCleanups) cleanup()
  watcherCleanups = []
}

/**
 * Subscribe to change events. Callback is invoked when watched files change (debounced).
 * @param {() => void} send - Callback to run on change
 * @returns {() => void} Unsubscribe function
 */
export function subscribe(send) {
  subscribers.add(send)
  return () => subscribers.delete(send)
}
