// bmad-dashboard/server/routes.js
import path from 'path'
import fs from 'fs'
import express from 'express'
import { packageRoot, isRunAsPackage, projectRoot, bmadRoot } from './config.js'
import { loadYaml, loadCsv } from './fileUtils.js'
import { getBmmConfig, detectArtifacts, inferBmmPhase, getSprintStatus, buildNextSteps, getLastExecutedArtifactKey } from './bmadState.js'
import { getArtifactDetails } from './artifactDetails.js'
import { subscribe } from './fileWatch.js'

export function registerRoutes(app) {
  app.get('/api/bmad-state/events', (req, res) => {
    res.setHeader('Content-Type', 'text/event-stream')
    res.setHeader('Cache-Control', 'no-cache')
    res.setHeader('Connection', 'keep-alive')
    res.flushHeaders?.()

    const unsubscribe = subscribe(() => {
      try {
        res.write('data: {"type":"change"}\n\n')
        res.flush?.()
      } catch {
        unsubscribe()
      }
    })

    req.on('close', () => unsubscribe())
  })

  app.get('/api/bmad-state', async (req, res) => {
    try {
      const manifest = (await loadYaml(path.join(bmadRoot, '_config', 'manifest.yaml'))) || {}
      const bmm = await getBmmConfig()
      const helpPath = path.join(bmadRoot, '_config', 'bmad-help.csv')
      const workflows = fs.existsSync(helpPath) ? await loadCsv(helpPath) : []
      const artifacts = await detectArtifacts(bmm)
      const phaseInfo = inferBmmPhase(artifacts, workflows)
      const sprintStatus = bmm.implementation_artifacts ? await getSprintStatus(bmm.implementation_artifacts) : null
      const nextSteps = buildNextSteps(workflows, phaseInfo, artifacts)
      const lastExecutedArtifactKey = getLastExecutedArtifactKey(bmm)
      let artifactDetails = null
      try {
        artifactDetails = await getArtifactDetails(projectRoot, bmm.planning_artifacts, bmm.implementation_artifacts, sprintStatus)
      } catch (e) {
        console.error('getArtifactDetails failed:', e)
      }

      const installedModules = (manifest.modules || []).map((m) => (m && m.name) ? m.name : '').filter(Boolean)
      const modulesToUse = installedModules.length > 0 ? installedModules : [...new Set(workflows.map((w) => w.module).filter(Boolean))]
      const allWorkflows = workflows
        .filter((w) => w.module && modulesToUse.includes(w.module))
        .map((w) => ({
          phase: w.phase,
          sequence: w.sequence,
          name: w.name,
          code: w.code,
          command: w.command,
          required: w.required === 'true',
          agent: w['agent-display-name'],
          agentTitle: w['agent-title'] || '',
          module: w.module || '',
          description: w.description,
        }))

      const bmmWorkflows = allWorkflows.filter((w) => w.module === 'bmm')

      res.json({
        projectRoot,
        projectName: bmm.project_name || 'Project',
        manifest: { installation: manifest.installation, modules: manifest.modules },
        bmm: { planningPath: bmm.planning_artifacts, implementationPath: bmm.implementation_artifacts },
        artifacts,
        artifactDetails,
        phase: phaseInfo.phase,
        nextSteps,
        sprintStatus,
        lastExecutedArtifactKey: lastExecutedArtifactKey || null,
        workflows: bmmWorkflows.filter((w) => w.phase !== 'anytime'),
        workflowsAll: bmmWorkflows,
        workflowsAllModules: allWorkflows,
      })
    } catch (err) {
      console.error(err)
      res.status(500).json({ error: 'Internal server error' })
    }
  })

  const distPath = path.join(packageRoot, 'dist')
  if (isRunAsPackage && fs.existsSync(distPath)) {
    app.use(express.static(distPath))
    app.get('*', (req, res) => {
      if (req.path.startsWith('/api')) return res.status(404).end()
      res.sendFile(path.join(distPath, 'index.html'))
    })
  }
}
